src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClientX/IO/TdsWriter.cs

```diff
@@ -0,0 +1,202 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Buffers.Binary;
+using System.Diagnostics;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Data.SqlClientX.IO
+{
+    /// <summary>
+    /// This class provides helper methods for writing bytes using <see cref="TdsStream"/>
+    /// It extends <see cref="TdsBufferAlloc"/> that manages allocations of bytes buffer for better memory management.
+    /// </summary>
+    internal class TdsWriter : IAsyncDisposable, IDisposable
+    {
+        private readonly TdsBufferAlloc _alloc;
+
+        /// <summary>
+        /// Instantiate TdsWriter with a TdsWriteStream
+        /// </summary>
+        /// <exception cref="ArgumentNullException"></exception>
+        public TdsWriter()
+        {
+            _alloc = new TdsBufferAlloc();
+        }
+
+        #region Public APIs
+
+        /// <summary>
+        /// Writes short value to out buffer, as little-endian.
+        /// </summary>
+        /// <param name="stream">Tds Stream to work with</param>
+        /// <param name="v">Value to write</param>
+        /// <param name="isAsync">Whether caller method is executing asynchronously.</param>
+        /// <param name="ct">Cancellation token</param>
+        public async ValueTask WriteShortAsync(TdsStream stream, short v, bool isAsync, CancellationToken ct)
+        {
+            var len = sizeof(short);
+            BinaryPrimitives.WriteInt16LittleEndian(_alloc.GetBuffer(len), v);
+            await WriteBytesAsync(stream, _alloc.GetBuffer(len), 0, len, isAsync, ct).ConfigureAwait(false);
+        }
+
+        /// <summary>
+        /// Writes int value to out buffer, as little-endian.
+        /// </summary>
+        /// <param name="stream">Tds Stream to work with</param>
+        /// <param name="v">Value to write</param>
+        /// <param name="isAsync">Whether caller method is executing asynchronously.</param>
+        /// <param name="ct">Cancellation token</param>
+        public async ValueTask WriteIntAsync(TdsStream stream, int v, bool isAsync, CancellationToken ct)
+        {
+            var len = sizeof(int);
+            BinaryPrimitives.WriteInt32LittleEndian(_alloc.GetBuffer(len), v);
+            await WriteBytesAsync(stream, _alloc.GetBuffer(len), 0, len, isAsync, ct).ConfigureAwait(false);
```

Comment:
Nit: in situations like this (where the async method call is the last in the method; the async method being called and the calling method both return `Task`/`ValueTask`, or the calling method simply returns the awaited value) we can simply return the ValueTask wholesale as below:

```suggestion
            return WriteBytesAsync(stream, _alloc.GetBuffer(len), 0, len, isAsync, ct);
```

I think this reduces the burden on the thread pool slightly in the async case. It also means that the method doesn't necessarily need to be declared async (not strictly a problem here.)